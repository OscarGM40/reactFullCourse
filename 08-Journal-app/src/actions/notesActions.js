import Swal from "sweetalert2";
import { fileUpload } from "../helpers/fileUpload";
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";
import { db } from "../firebase/firebase-config";

// podriamos pasar el uid de cada usuario que tiene en Firebase como argumento pero también podemos usar una callback que nos proporciona thunk y nos devuelve el state.El nombre el que quiera,es el segundo argumento del return(getState aqui)
export const startNewNote = () => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid;
    
    const newNote = {
      title: "",
      body: "",
      date: new Date().getTime(), //mls actuales
    };
    
    try {
      /* Desde firebase 9 el método es addDoc y no add */
      const docRef = await db.collection(`${uid}/journal/notes`).add(newNote);
      // console.log(docRef);
      dispatch(activateNote(docRef.id,newNote))
      dispatch(addNewNote(docRef.id,newNote))
      
    } catch (error) {
      console.log(error);
    }
  };
};

export const addNewNote = (id,note) => ({
  type: types.notesAddNew,
  payload: {
    id,
    ...note,
  },
});

export const startSaveNote = ( note ) => {
  return async(dispatch,getState) => {
    // fijate que necesito el uid del usuario luego uso la cb
    const uid = getState().auth.uid;

    //  si no viene el url lo borro pues da error el undefined
    /* if(!note.url){
      delete note.url;
    } */
    
    // en realidad no necesito el delete,puedo cambiar el undefined por un null ya que puedo guardar un null  
    // saco las propiedades
     const { id, url=null,...rest} = note;
    //  console.log(url)
     /*  console.log(note,'note')
      console.log(id,'id') */

    /* const noteToFirestore = { ...note };
    delete noteToFirestore.id; */

    try {
      // db.collection es para trabajar en una coleccion
      // await db.collection(`path...`)
      // lo que quiero ahora es trabajar en un documento,
      await db.doc(`${uid}/journal/notes/${note.id}`)
      .update({ url, ...rest })  
      dispatch(refreshNote(note))
      Swal.fire("Saved",rest.title, 'success');
    } catch (error) {
      console.log(error)
    }

  }  
}

export const refreshNote = (note) => ({
  type:types.notesUpdate,
  payload: note
})

export const activateNote = (id,note) => ({
  type: types.notesActive,
  payload: { id, ...note }
}) // <- fijate que de esta forma las acciones sincronas devuelven un objeto luego return => ({})


export const startLoadingNotes = (uid) => {
  return async (dispatch) => {
    const notes = await loadNotes(uid)
    dispatch(setNotes(notes))
  }
}

// no haría falta romper la referencia porque estoy creando un nuevo arreglo y no apunto al de el state!
export const setNotes = (notes) => ({
  type:types.notesLoad,
  payload: notes
})

// como es una tarea asincrona es obvio qué debo usar
export const startUploading = (file) => {
  return async ( dispatch, getState ) => {
    const { active: activeNote } = getState().notes;

    Swal.fire({ 
      title: 'Uploading...',
      text: 'Please wait...',
      allowOutsideClick: false,
      onBeforeOpen: () => {
        Swal.showLoading()
      }
    });
    // llamo a mi helper que me devuelve el campo secure_url
    const fileUrl = await fileUpload(file);
    activeNote.url = fileUrl;
    // console.log(fileUrl,'fileUrl')

    dispatch(startSaveNote(activeNote))
    Swal.close();
    
  }
}

export const startDeleteNote = (id) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid;
    
    try {
      await db.doc(`${uid}/journal/notes/${id}`).delete();
      dispatch(deleteNote(id));
    } catch (error) {
      console.log(error);
    }
  }
}

export const deleteNote = (id) => ({
  type: types.notesDelete,
  payload: id
})

export const notesLogout = () => ({
  type: types.notesLogoutCleaning
})

