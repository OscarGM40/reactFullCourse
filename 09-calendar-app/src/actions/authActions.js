import { fetchSinToken } from "../helpers/fetch";
import { types } from "../types/types";
import Swal from "sweetalert2";

export const startLogin = (email, password) => {
  return async (dispatch, getState) => {
    const response = await fetchSinToken(
      "auth",
      {
        email,
        password,
      },
      "POST"
    );
    const body = await response.json();
    
    if(body.ok){
      localStorage.setItem("token", body.token);
      localStorage.setItem("token-init-date", new Date().getTime());

      dispatch(login({
        uid:body.uid,
        name:body.name,
      }));
    // si no viene el ok a true
    } else {
      Swal.fire('Error', body.msg, 'error');
    }
  };
};

const login = (user) => ({
  type: types.authLoginSuccess,
  payload: user,
})

export  const startRegister = ( name,email,password ) => {
  return async (dispatch, getState) => {
    const response = await fetchSinToken(
      "auth/new",
      {
        name,
        email,
        password,
      },
      "POST"
    );
    const body = await response.json();
   
    if(body.ok){
      localStorage.setItem("token", body.token);
      localStorage.setItem("token-init-date", new Date().getTime());
      dispatch(login({
        uid:body.uid,
        name:body.name,
      }));
    } else {
      Swal.fire('Error', body.msg, 'error');
    }
  }
}