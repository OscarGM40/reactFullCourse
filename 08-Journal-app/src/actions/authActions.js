import { firebase, googleAuthProvider } from "../firebase/firebase-config";
import { types } from "../types/types";
import { startLoading, finishLoading } from "../actions/uiActions";
import Swal from "sweetalert2";
import { notesLogout } from "./notesActions";

// fijate que algunos si tienen la lógica(los asincronos) mientras que los sincronos( no sé si son los de redux también) solo declaran la accion.Es un poco confusa esta forma.
export const startLoginEmailPassword = (email, password) => {
  return async (dispatch) => {
    dispatch(startLoading());

    try {
      const { user } = await firebase
        .auth()
        .signInWithEmailAndPassword(email, password);
      dispatch(login(user.uid, user.displayName));
      dispatch(finishLoading());
    } catch (e) {
      console.log("error en el login");
      dispatch(finishLoading());
      Swal.fire("Error", e.message, "error");
      return e.code;
    }
  };
};

export const startRegisterWithEmailPasswordName = (email, password, name) => {
  return async (dispatch) => {
    // firebase trabajará con promesas,ojo
    try {
      const userCred = await firebase
        .auth()
        .createUserWithEmailAndPassword(email, password);
      const { user } = userCred;
      // es muy conveniente acceder a este método y subir el nombre también
      await user.updateProfile({ displayName: name });
      // console.log(user)
      dispatch(login(user.uid, user.displayName));
    } catch (e) {
      console.log("error en el registro");
      Swal.fire("Error", e.message, "error");
      return e.code;
    }
  };
};

export const startGoogleLogin = () => {
  // con esta forma thunk provee del dispatch en el return,es por esto que puedo usarlo
  //recuerda que esta forma con el return es por usar redux y no @reduxjs/toolkit.No es díficil,sólo cambia un poco
  return (dispatch) => {
    //singInWithPopup pide un proveedor y es una promesa
    firebase
      .auth()
      .signInWithPopup(googleAuthProvider)
      // la función resuelve a un UserCredential
      .then((userCred) => {
        const { user } = userCred;
        dispatch(login(user.uid, user.displayName));
      });
  };
};

export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName,
  },
});

// si bien es un poquito más liosa esta forma también parece interesante.
export const startLogout = () => {
  // es asincrono porque hay comunicacion con un backend asin
  // fijate que puedo usar async en el return y trabajar con await en el interior.Realmente return sólo indica que es una instrucción terminal.
  return async (dispatch) => {
    // si bien es una Promise no se suele usar el catch pues no suele fallar
    await firebase.auth().signOut();
    dispatch(logout());
    dispatch(notesLogout());
  };
};

export const logout = () => ({
  type: types.logout,
});
