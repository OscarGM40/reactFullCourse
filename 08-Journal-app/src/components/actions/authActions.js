import { firebase, googleAuthProvider } from "../firebase/firebase-config";
import { types } from "../../types/types";
import { startLoading, finishLoading } from "../actions/uiActions";
import Swal from "sweetalert2";

// fijate que algunos si tienen la lógica(los asincronos) mientras que los sincronos( no sé si son los de redux también) solo declaran la accion.Es un poco confusa esta forma.
export const startLoginEmailPassword = (email, password) => {
  return (dispatch) => {
    dispatch(startLoading());
    firebase
      .auth()
      .signInWithEmailAndPassword(email, password)
      .then(({ user }) => {
        dispatch(login(user.id, user.displayName));
        dispatch(finishLoading());
      })
      .catch((e) => {
        console.log('por aqui');
        dispatch(finishLoading());
        Swal.fire("Error", e.message, "error");
      });
  };
};

export const startRegisterWithEmailPasswordName = (email, password, name) => {
  return (dispatch) => {
    // firebase trabajará con promesas,ojo
    firebase
      .auth()
      .createUserWithEmailAndPassword(email, password)
      .then(async (userCred) => {
        const { user } = userCred;
        // es muy conveniente acceder a este método y subir el nombre también
        await user.updateProfile({ displayName: name });
        // console.log(user)
        dispatch(login(user.id, user.displayName));
      })
      .catch((e) => {
        console.log(e);
        Swal.fire("Error", e.message, "error");
      });
    };
  };
  
  export const startGoogleLogin = () => {
    // con esta forma thunk provee del dispatch en el return,es por esto que puedo usarlo
    //recuerda que esta forma con el return es por usar redux y no @reduxjs/toolkit.No es díficil,sólo cambia un poco
    return (dispatch) => {
      //singInWithPopup pide un proveedor y es una promesa
      firebase
      .auth()
      .signInWithPopup(googleAuthProvider)
      // la función resuelve a un UserCredential
      .then((userCred) => {
        const { user } = userCred;
        dispatch(login(user.uid, user.displayName));
      });
    };
  };
  
  export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
      uid,
      displayName,
    },
  });
  
  // si bien es un poquito más liosa esta forma también parece interesante.
  export const startLogout = () => {
    // es asincrono porque hay comunicacion con un backend asin
    // fijate que puedo usar async en el return y trabajar con await en el interior.Realmente return sólo indica que es una instrucción terminal.
    return async (dispatch) => {
      // si bien es una Promise no se suele usar el catch pues no suele fallar
      await firebase.auth().signOut();
      dispatch(logout);
    };
  };
  
  export const logout = () => ({
    type: types.logout,
  });
  