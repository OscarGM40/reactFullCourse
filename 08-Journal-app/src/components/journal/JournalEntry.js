import React from "react";
import moment from "moment";
// si quiero la traduccion importar esto ↓↓
// import "moment/locale/es";
import { useDispatch } from "react-redux";
import { activateNote } from "../actions/notesActions"



export const JournalEntry = ({ id, title, body, date,url }) => {

  const dispatch = useDispatch()
  const noteDate = moment(date);

  const setActiveNote = () => {
      dispatch(activateNote(id,{title,body,date,url}));
  }
  
  
  return (
    <div className="journal__entry pointer animate__animated animate__flipInY animate__fast"
      onClick={setActiveNote}>
      {
        url &&
        <div
          className="journal__entry-picture"
          style={{
            backgroundSize: "cover",
            backgroundImage: `url(${url})`,
          }}
        >
        </div>
      }
      <div className="journal__entry-body">
        <p className="journal__entry-title">{title}</p>
        <p className="journal__entry-content">{body}</p>
      </div>

      <div className="journal__entry-date-box">
        <span>{ noteDate.format("dddd") }</span>
        <h4>{ noteDate.format("Do") }</h4>
      </div>
    </div>
  );
};
