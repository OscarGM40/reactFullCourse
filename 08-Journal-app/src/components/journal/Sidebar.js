import React from "react";
import { useSelector } from "react-redux";
import { useDispatch } from "react-redux";
import { startLogout } from "../../actions/authActions";
import { startNewNote } from "../../actions/notesActions";
import { JournalEntries } from "./JournalEntries";

export const Sidebar = () => {
  

  const dispatch = useDispatch()
  const { name } = useSelector( state => state.auth) 

  // fijate que con esta forma disparo aqui la accion y en el otro file este método a su vez también dispara tras la asincronia el borrado de redux.Interesante,realmente puede programarse lo que se quiera
  const handleLogout = () => {
    dispatch(startLogout());
    // setTimeout(() => window.location.href="/auth/login",300)
  }

  const handleAddNew = () => {
    dispatch(startNewNote());
  }
  
  return (
    <aside className="journal__sidebar">
      <div className="journal__sidebar-navbar">
        <h3 className="mt-5">
          <i className="far fa-moon"></i>
          <span> {name}</span>
        </h3>

        <button className="btn"
          onClick={handleLogout}>Logout</button>
      </div>

      <div className="journal__new-entry"
        onClick={handleAddNew}>
        <i className="far fa-calendar-plus fa-5x "></i>
        <p className="mt-5">New Entry</p>
      </div>

      <JournalEntries />
      
    </aside>
  );
};
