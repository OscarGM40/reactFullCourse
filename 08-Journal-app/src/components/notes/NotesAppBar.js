import React, { useRef } from "react";
import { useDispatch, useSelector } from "react-redux";
import { startSaveNote, startUploading } from "../../actions/notesActions";
import moment from "moment";
import "moment/locale/es"

export const NotesAppBar = () => {
   
  const dispatch = useDispatch();
  const { active } = useSelector((state) => state.notes);
  const inputRef = useRef();
  // console.log(active,'active')

  const handleSave = (e) => {
    e.preventDefault()
    dispatch(startSaveNote(active));
  };

  const handlePicture = (e) => {
   //   no necesito ni siquiera una label,solo simular el click,pero que grande eres Fernando
   //  document.querySelector("#fileSelector").click();
   inputRef.current.click()
  }

  const handleFile = (e) => {
   const file = e.target.files[0];
   if(file){
      dispatch(startUploading(file))
   }
  }

  return (
    <div className="notes__appbar">
      <span>{moment().format("LL")}</span>


      <input 
      //   id="fileSelector"
        ref={inputRef}
        type="file"
        style={{display:"none"}}
        onChange={(e) => handleFile(e)} />


      <div className="">
        <button onClick={handlePicture} className="btn">
          Picture
        </button>
        <button onClick={handleSave} className="btn">
          Save
        </button>
      </div>
    </div>
  );
};
