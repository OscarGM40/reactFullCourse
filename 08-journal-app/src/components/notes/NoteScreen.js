import React from 'react'
import { NotesAppBar } from './NotesAppBar'

export const NoteScreen = () => {
   return (
      <div className="notes__main-content">
         
          <NotesAppBar />

          <div className="notes__content">

            <input 
            type="text"
            placeholder="Some awesome title"
            className="notes__title-input"
            autoComplete="off"
            />

            <textarea name="" cols="30" rows="2"
            placeholder="What happened today"
            className="notes__textarea"
            >
            </textarea>

               <div className="notes__image"  style={{
          backgroundSize: "cover",
          backgroundImage: "url(https://picsum.photos/400/250?random=2)",
        }}>
                  {/* <img src="https://picsum.photos/450/200?random=2" alt="imagen"/> */}
               </div>

          </div>


      </div>
   )
}
