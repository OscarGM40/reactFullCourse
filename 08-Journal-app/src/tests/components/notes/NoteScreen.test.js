import React from "react";
import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { MemoryRouter } from "react-router-dom";
import { Sidebar } from "../../../components/journal/Sidebar";

import { activateNote } from "../../../actions/notesActions";
import { NoteScreen } from "../../../components/notes/NoteScreen";

jest.mock("../../../actions/notesActions", () => ({
  activateNote: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initialState = {
  auth: {
    uid: "123",
    name: "John",
  },
  ui: {
    loading: false,
    msgError: null,
  },
  notes: {
    active: {
      id: 1234,
      title: "hola",
      body: "mundo",
    },
    notes: [],
  },
};

let store = mockStore(initialState);
store.dispatch = jest.fn();

const wrapper = mount(
  <MemoryRouter>
    <Provider store={store}>
      <NoteScreen />
    </Provider>
  </MemoryRouter>
);

describe("Pruebas en el NoteScreen", () => {
  beforeEach(() => {
    store = mockStore(initialState);
    store.clearActions();
  });

  test("debe de mostrarse correctamente", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("debe de llamar el activateNote al cambiar el formulario", async () => {
    wrapper.find("input[name='title']").simulate("change", {
      target: { name: "title", value: "adios" },
    });
    expect(activateNote).toHaveBeenCalled();
    expect(activateNote).toHaveBeenLastCalledWith(1234, {
      title: "adios",
      id:1234,
      body: "mundo",
    });

  });
});
