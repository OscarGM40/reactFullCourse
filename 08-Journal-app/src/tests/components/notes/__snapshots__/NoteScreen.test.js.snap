// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el NoteScreen debe de mostrarse correctamente 1`] = `
<div
  className="notes__main-content"
>
  <div
    className="notes__appbar"
  >
    <span>
      25 de noviembre de 2021
    </span>
    <input
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div
      className=""
    >
      <button
        className="btn"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      autoComplete="off"
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Some awesome title"
      type="text"
      value="hola"
    />
    <textarea
      className="notes__textarea"
      cols="30"
      name="body"
      onChange={[Function]}
      placeholder="What happened today"
      rows="2"
      value="mundo"
    />
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
